iceoryx_posh/source/mepoo/chunk_header.cpp

```diff
@@ -23,9 +23,24 @@ ChunkHeader::ChunkHeader() noexcept
 {
 }
 
-ChunkHeader* convertPayloadPointerToChunkHeader(const void* const payload) noexcept
+void* ChunkHeader::payload() const noexcept
 {
-    return reinterpret_cast<ChunkHeader*>(reinterpret_cast<uint64_t>(payload) - sizeof(ChunkHeader));
+    // payload is always located relative to "this" in this way
+    return reinterpret_cast<void*>(reinterpret_cast<uint64_t>(this) + m_payloadOffset);
+}
+
+ChunkHeader* ChunkHeader::fromPayload(const void* const payload) noexcept
+{
+    if (payload == nullptr)
+    {
+        return nullptr;
+    }
+    auto payloadAddress = reinterpret_cast<uint64_t>(payload);
+    // the payload offset is always stored in front of the payload, no matter if a custom header is used or not
+    // or if the payload has a custom allignment
+    using PayloadOffsetType = decltype(ChunkHeader::m_payloadOffset);
+    auto payloadOffset = reinterpret_cast<PayloadOffsetType*>(payloadAddress - sizeof(PayloadOffsetType));
+    return reinterpret_cast<ChunkHeader*>(payloadAddress - *payloadOffset);
```

Comment:
What happens when `payloadAddress < *payloadOffset`. Then we would have an underflow and would return maybe not what we want?! I know that this is a corner case but I think we should cover it somehow since this case can maybe occur when we have a programming error on a higher level and exactly at this point the whole system gets screwed up.

Or we try to avoid this issue by changing the parameter since this is at the end public API?!

Reply:
In this case you're screwed. Although this is public API, it should only be used in untyped API and I think with the new API it might possible to fully get rid of this, except maybe for the C API. The issue is not just when `payloadAddress < *payloadOffset`. Since `payloadOffset` will probably never be larger than a few hundred bytes, this is quite a corner case. The far bigger problem is that the payload pointer can be anything, since it's a void pointer. We also cannot do any sanity checks here. It's just a "we give you a ChunkHeader pointer if you promise to give us a payload pointer"

Reply:
Often we release a chunk after the `fromPayload` call. If this is then not a valid chunk the user will get a nice cup of error handler